std/algebra/emulated/sw_bn254/pairing.go

```diff
@@ -644,12 +642,150 @@ func (pr Pairing) MillerLoopAndMul(P *G1Affine, Q *G2Affine, previous *GTEl) (*G
+			B2: fields_bn254.E2{A0: *hint[28], A1: *hint[29]},
+		},
+		C1: fields_bn254.E6{
+			B0: fields_bn254.E2{A0: *hint[30], A1: *hint[31]},
+			B1: fields_bn254.E2{A0: *hint[32], A1: *hint[33]},
+			B2: fields_bn254.E2{A0: *hint[34], A1: *hint[35]},
+		},
+	}
+
+	if Q.Lines == nil {
+		Qlines := pr.computeLines(&Q.P)
+		Q.Lines = &Qlines
+	}
+	lines := *Q.Lines
+
+	// precomputations
+	yInv := pr.curveF.Inverse(&P.Y)
+	xNegOverY := pr.curveF.MulMod(&P.X, yInv)
+	xNegOverY = pr.curveF.Neg(xNegOverY)
+
+	// init Miller loop accumulator to residueWitnessInv to share the squarings
+	// of residueWitnessInv^{6x₀+2}
+	res := &residueWitnessInv
+
+	// Compute f_{6x₀+2,Q}(P)
+	for i := 64; i >= 0; i-- {
+		res = pr.Square(res)
+
+		switch loopCounter[i] {
+		case 0:
+			// ℓ × res
+			res = pr.MulBy034(
+				res,
+				pr.MulByElement(&lines[0][i].R0, xNegOverY),
+				pr.MulByElement(&lines[0][i].R1, yInv),
+			)
+		case 1:
+			// multiply by residueWitnessInv when bit=1
+			res = pr.Mul(res, &residueWitnessInv)
+			// lines evaluations at P
+			// and ℓ × ℓ
+			prodLines := pr.Mul034By034(
+				pr.MulByElement(&lines[0][i].R0, xNegOverY),
+				pr.MulByElement(&lines[0][i].R1, yInv),
+				pr.MulByElement(&lines[1][i].R0, xNegOverY),
+				pr.MulByElement(&lines[1][i].R1, yInv),
+			)
+			// (ℓ × ℓ) × res
+			res = pr.MulBy01234(res, prodLines)
+		case -1:
+			// multiply by 1/residueWitnessInv=residueWitness when bit=-1
+			res = pr.Mul(res, &residueWitness)
+			// lines evaluations at P
+			// and ℓ × ℓ
+			prodLines := pr.Mul034By034(
+				pr.MulByElement(&lines[0][i].R0, xNegOverY),
+				pr.MulByElement(&lines[0][i].R1, yInv),
+				pr.MulByElement(&lines[1][i].R0, xNegOverY),
+				pr.MulByElement(&lines[1][i].R1, yInv),
+			)
+			// (ℓ × ℓ) × res
+			res = pr.MulBy01234(res, prodLines)
+		default:
+			return nil
+		}
+	}
+
+	// Compute  ℓ_{[6x₀+2]Q,π(Q)}(P) · ℓ_{[6x₀+2]Q+π(Q),-π²(Q)}(P)
+	// lines evaluations at P
+	// and ℓ × ℓ
+	prodLines := pr.Mul034By034(
+		pr.MulByElement(&lines[0][65].R0, xNegOverY),
+		pr.MulByElement(&lines[0][65].R1, yInv),
+		pr.MulByElement(&lines[1][65].R0, xNegOverY),
+		pr.MulByElement(&lines[1][65].R1, yInv),
+	)
+	// (ℓ × ℓ) × res
+	res = pr.MulBy01234(res, prodLines)
+
+	// multiply by previous multi-Miller function
+	res = pr.Mul(res, previous)
+
+	// Check that  res * cubicNonResiduePower * residueWitnessInv^λ' == 1
+	// where λ' = q^3 - q^2 + q, with u the BN254 seed
+	// and residueWitnessInv, cubicNonResiduePower from the hint.
+	// Note that res is already MillerLoop(P,Q) * residueWitnessInv^{6x₀+2} since
+	// we initialized the Miller loop accumulator with residueWitnessInv.
+	t2 := pr.Mul(&cubicNonResiduePower, res)
+
+	t1 := pr.FrobeniusCube(&residueWitnessInv)
+	t0 := pr.FrobeniusSquare(&residueWitnessInv)
+	t1 = pr.DivUnchecked(t1, t0)
+	t0 = pr.Frobenius(&residueWitnessInv)
+	t1 = pr.Mul(t1, t0)
+
+	t2 = pr.Mul(t2, t1)
+
+	pr.AssertIsEqual(t2, pr.One())
+
```

Comment:
Doesn't it lack a check that mul(residue_witness, residue_witness_inv) == 1 ?

Reply:
yeah you're right. Either this or to compute the inverse in-circuit.